R/quosure.R

```diff
@@ -0,0 +1,166 @@
+#' Is an object a quosure or quosure-like?
+#'
+#' @description
+#'
+#' These predicates test for [quosure][quosure] objects.
+#'
+#' - `is_quosure()` tests for the canonical R quosure: the one-sided
+#'   "formula".
+#'
+#' - `is_quosureish()` tests for general R quosure objects: quosures,
+#'   two-sided formulas, and [definitions][op-definition].
+#'
+#'
+#' @details
+#'
+#' The `scoped` argument patterns-match on whether the scoped bundled
+#' with the quosure is valid or not. Invalid scopes may happen in
+#' nested quotations like `~~expr`, where the outer quosure is validly
+#' scoped but not the inner one. This is because `~` saves the
+#' environment when it is evaluated, and quoted quosures are by
+#' definition not evaluated. Note that in the [tidy evaluation
+#' framework][tidy_eval], unscoped quosures are automatically given
+#' the environment of the outer quosure during the evaluation process.
+#'
+#' @param x An object to test.
+#' @param scoped Whether the quosure is scoped, that is, has a valid
+#'   environment attribute. If `NULL`, the quosure scope is not
+#'   inspected.
+#' @seealso [as_quosure()][quosure] and [quosure()] for creating
+#'   quosures, and [tidy_quote()] or [tidy_eval()] for information
+#'   about the role of quosures in the tidy evaluation framework.
+#' @export
+#' @examples
+#' # Degenerate quosures are often created by quoting, since `~`
+#' # records the environment when it is evaluated the first time:
+#' f <- ~~expr
+#'
+#' # The outer quosure has been evaluated and is scoped:
+#' is_quosure(f, scoped = TRUE)
+#'
+#' # But the inner formula is not:
+#' inner_f <- f_rhs(f)
+#' is_quosure(inner_f, scoped = TRUE)
+#'
+#'
+#' # Formulas and definitions are not quosures:
+#' is_quosure(a := b)
+#' is_quosure(a ~ b)
+#'
+#' # But they are quosureish objects:
+#' is_quosureish(a := b)
+#' is_quosureish(a ~ b)
+#' @md
```

Comment:
Yes I was planning to port the whole codebase to markdown once all branches are merged.